util/logmsg.c

```diff
@@ -84,11 +84,32 @@ static char *logmsg_level_str(int lvl)
     }
 }
 
-static int logmsgv_lk(loglvl lvl, const char *fmt, va_list args)
+static int sprintf_auto_resize(char **headp, int *offp, const char *fmt, const char *content)
+{
+    char *head = *headp;
+    int off = *offp;
+    char dummy[1];
+    int size = snprintf(dummy, 1, fmt, content);
+
+    head = realloc(head, off + size + 1);
```

Comment:
Given the log messages won't be too long, I think it's better to start with a static buffer of some decent size and fallback to malloc only for ones that won't fit.